### PR TITLE
Move JDBC warnings to Statement

### PR DESCRIPTION
**Pull request by electrum**
The warnings are not specific to the ResultSet.

### CODE REVIEW

File: presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java [hunk 1]
```diff
@@ -96,6 +98,16 @@ public void setup()
     {
         connection = createConnection();
         statement = connection.createStatement();
+
+        statement.execute("CREATE SCHEMA IF NOT EXISTS blackhole.blackhole");
+        statement.executeUpdate("" +
+                "CREATE TABLE IF NOT EXISTS blackhole.blackhole.slow_table (x int) " +
```
Comment (findepi) [inline presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java]: There is no need to try-recreate this table per test method. Why not run this in `@BeforeClass`?

File: presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java [hunk 2]
```diff
@@ -122,69 +134,47 @@ public void testStatementWarnings()
...
     public void testLongRunningStatement()
-            throws SQLException, InterruptedException
+            throws Exception
     {
         ExecutorService queryExecutor = newSingleThreadExecutor(daemonThreadsNamed("test-%s"));
         QueryCreationFuture queryCreationFuture = new QueryCreationFuture();
         queryExecutor.submit(() -> {
             try {
-                statement.execute("CREATE SCHEMA blackhole.blackhole");
-                statement.execute("CREATE TABLE blackhole.blackhole.test_table AS SELECT 1 AS col1 FROM tpch.sf1.lineitem CROSS JOIN tpch.sf1.lineitem");
```
Comment (findepi) [inline presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java]: The refactor of this test that replaces lineitem self CROSS JOIN with backhole slow table is not related to moving JDBC warnings to Statement and so should go in separate commit.

File: presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java [hunk 3]
```diff
@@ -122,69 +134,47 @@ public void testStatementWarnings()
...
-                // Wait for new warnings
-            }
-            warning = warning.getNextWarning();
-            assertTrue(currentWarnings.add(new WarningEntry(warning)));
-            Thread.sleep(100);
-        }
-        assertEquals(currentWarnings.size(), 100);
+
+        Set<WarningEntry> currentWarnings = getStatementWarnings(statement, queryCreationFuture);
+        assertThat(currentWarnings).size().isGreaterThanOrEqualTo(PRELOADED_WARNINGS + 1);
```
Comment (findepi) [inline presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java]: `assertThat(currentWarnings).size().isGreaterThan(PRELOADED_WARNINGS);`

File: presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java [hunk 4]
```diff
@@ -225,6 +215,30 @@ public void testSqlWarning()
         assertWarningsEqual(warning.getNextWarning().getNextWarning(), toPrestoSqlWarning(warnings.get(2)));
     }
 
+    private static Set<WarningEntry> getStatementWarnings(Statement statement, Future<?> future)
+            throws Exception
+    {
+        Set<WarningEntry> warnings = new HashSet<>();
+        SQLWarning warning = null;
+        while (!future.isDone()) {
+            SQLWarning nextWarning;
```
Comment (findepi) [inline presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java]: Technically you don't  need two variables: `warning` and `nextWarning`. One would do.

File: presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java (hunk 3 above)
Comment (findepi) [reply presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java]: Previously we were checking there were 100 warnings, not we're checking there are at least 6.
I don't know where this 100 came from, but in any case, it would be best to check actual messages of these warnings.

File: presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java [hunk 5]
```diff
@@ -122,69 +134,47 @@ public void testStatementWarnings()
...
-            else {
-                break;
-            }
-            Thread.sleep(100);
-        }
+
+        Set<WarningEntry> currentWarnings = getStatementWarnings(statement, queryCreationFuture);
+        assertThat(currentWarnings).size().isGreaterThanOrEqualTo(PRELOADED_WARNINGS + 1);
+
         queryExecutor.shutdownNow();
```
Comment (findepi) [inline presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java]: This should be `finally`

File: presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoResultSet.java [hunk 6]
```diff
@@ -1761,35 +1758,17 @@ private static Number toNumber(Object value)
...
-            }
-            return updateType == null;
         }
 
         @Override
         protected Iterable<List<Object>> computeNext()
         {
-            if (isQuery) {
-                // Clear the warnings if this is a query, per ResultSet javadoc
-                warningsManager.clearWarnings();
```
Comment (findepi) [inline presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoResultSet.java]: Currently the test pulls everything into single bag of warnings (with `getStatementWarnings` method) and so we're not testing when warnings are visible and when they get cleared.

Would be good to have explicit test for the tested behavior. Like:

```
try (statement = ..) {
   try (rs = ...) {
      rs.next();
      check statement.getWarnings
      check rs.getWarnings
      rs.next();
      check statement.getWarnings
      check rs.getWarnings
  }
}
```

File: presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java (hunk 3 above)
Comment (electrum) [reply presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java]: The messages are not useful, as they are generated in `TestingWarningCollector` by rendering the warning code. However, the code and message are compared for uniqueness via the `WarningEntry` wrapper which is then inserted into a `Set`.

The previous test code was confusing and didn't really test what we wanted, which is to ensure that warnings show up as the query is running. I rewrote the code to directly test that.

File: presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java (hunk 5 above)
Comment (electrum) [reply presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java]: Agreed, though I think that makes the code harder to read. I'll move it to the setup/tearndown.

File: presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoResultSet.java (hunk 6 above)
Comment (electrum) [reply presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoResultSet.java]: Agreed, see comment above

**Comment by electrum** [comment]
@findepi thanks for the review. I updated/rewrote the PR

File: presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java [hunk 7]
```diff
@@ -343,27 +330,4 @@ public int hashCode()
             return Objects.hash(vendorCode, sqlState, message);
         }
     }
-
-    private static class QueryCreationFuture
-            extends AbstractFuture<QueryInfo>
```
Comment (findepi) [inline presto-jdbc/src/test/java/io/prestosql/jdbc/TestJdbcWarnings.java]: yikes
